docker/webapp/app.py

```diff
@@ -0,0 +1,258 @@
+
+def maybe_duration(t):
+    if t is not None: return datetime.timedelta(seconds=t)
+
+class PackageData:
+    def __init__(self):
+        # Load raw data
+        with open(os.path.join(HTML_DIR, "recipes.json"), 'r') as f:
+            _recipes = json.load(f)
+        with open(os.path.join(HTML_DIR, "archive.json"), 'r') as f:
+            _archive = json.load(f)
+        with open(os.path.join(HTML_DIR, "download_counts.json"), 'r') as f:
+            _download_counts = json.load(f)
+        with open(os.path.join(HTML_DIR, "build-status.json"), 'r') as f:
+            _build_status = json.load(f)
+
+        # Assemble package info
+        self.packages_by_name = {name: PackageDescriptor(name, entry=entry,
+                              recipe=_recipes[name],
+                              download_counts=_download_counts)
+            for (name, entry) in _archive.items()}
+
+        self.packages = sorted(self.packages_by_name.values(),
+                               key=lambda p: p.name) # Pre-sort by the default case
+
+        # Pre-calculate download percentiles
+        by_downloads = sorted(self.packages, key=lambda p: p.downloads)
+        percentile = 0
+        self.download_percentiles = {}
+        for (_, cohort) in groupby(by_downloads, lambda p: p.downloads):
+            cohort = list(cohort)
+            self.download_percentiles.update({ p.name: percentile / len(by_downloads) for p in cohort })
+            percentile += len(cohort) * 100
+
+        self.total_downloads = sum(_download_counts.values())
+        self.last_build = BuildStatus(started_at=maybe_timestamp(_build_status["started"]),
+                                      completed_at=maybe_timestamp(_build_status["completed"]),
+                                      next_at=maybe_timestamp(_build_status["next"]),
+                                      duration=maybe_duration(_build_status["duration"]))
+    def is_stale(self):
+        return False
+
+package_data = None
+def load_package_data():
+    global package_data
+    if package_data is None or package_data.is_stale():
+        package_data = PackageData()
+    return package_data
+
+##############################################################################
+# Pagination
+##############################################################################
+class Paginator:
+    def __init__(self, items, page_size, page):
+        items = list(items)
+        self.page_size = page_size
+        self.first_page = 1
+        self.last_page = ceil(len(items) / page_size)
+        self.page_numbers = range(max(page - 5, 1), min(page + 5, self.last_page))
+
+        if page < self.first_page or page > self.last_page:
+            raise IndexError("No such page")
+        start = (page - 1) * page_size
+        self.page_items = items[start:start+self.page_size]
+        self.page_number = page
+
+##############################################################################
+# Flask app
```

Comment:
Seems like an old comment/approach with Flask and then a change to FastAPI due to performance? If so, maybe consider using [Falcon](https://falconframework.org/) (benchmarks: [1](https://falconframework.org/#sectionBenchmarks), [2](https://gist.github.com/nhymxu/814cf9b3294276629d2231248b709e26)) especially PyPy-based one.

Reply:
Thanks, that looks good too. And yes, started with Flask, then felt like trying FastAPI too. I hadn't come across Falcon!